feat(managers): Add save to .csv file

ds-skoropad opened the pull request:
FileBackedTaskManager for the ability to save a list of tasks

Sprint 7 review 1

ds-skoropad committed:
feat(managers): Add save to .csv file

FileBackedTaskManager for the ability to save a list of tasks

Sprint 7 review 1

ds-skoropad committed:
fix: make SEPARATOR_CSV a constant

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,166 @@
+        return task;
+    }
+
+    private static class ManagerSaveException extends RuntimeException {
+        public ManagerSaveException(String message) {
+            super(message);
+        }
+    }
```
Вынеси исключение в отдельный класс, не очень хорошо все городит ьв одном классе)

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,166 @@
+        }
+    }
+
+    // В ТЗ написано создать данный метод статичным ... Зачем, почему не ясно ... Ответа в пачке я пока не получил...
```
Чтобы вызывать его без создания экземпляра класса и он еще должен создавать и возвращать восстановленный из файла менеджер, как пример:
`FileBackedTaskManager backedManager = FileBackedTaskManager.loadFromFile(file);` 
На выходе получаем новый экземпляр менеджера, восстановленный из файла

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,166 @@
+     * чтобы писать в него операции изменения, а не перезаписывать файл полностью каждый раз.
+     */
+    public void save() {
+        try (BufferedWriter writer = new BufferedWriter(new FileWriter(file.toFile(), StandardCharsets.UTF_8))) {
+            for (Task task : getTaskGroup()) {
+                writer.write(taskToStringCsv(task));
+                writer.newLine();
+            }
+            for (Task epic : getEpicTaskGroup()) {
+                writer.write(taskToStringCsv(epic));
+                writer.newLine();
+            }
+            for (Task sub : getSubTaskGroup()) {  // Важно подзадачи после эпиков записать
+                writer.write(taskToStringCsv(sub));
+                writer.newLine();
+            }
+        } catch (IOException e) {
+            throw new ManagerSaveException(e.getMessage());
+        }
```
Перед списком задач/подзадач/эпиков должен быть хедер файла с названиями полей

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,166 @@
+    }
+
+    // Column: id,type,name,status,description,epic
+    private String taskToStringCsv(Task task) {
+        if (task == null) return null;
+
+        String epicId = "";
+        if (task.getType() == TypeTask.SUB_TASK) {
+            epicId = Integer.toString(((SubTask) task).getEpicTaskId());
+        }
+        return String.join(SEPARATOR_CSV, Integer.toString(task.getId()), task.getType().toString(), task.getTitle(),
+                task.getStatus().toString(), task.getDescription(), epicId);
+    }
+
+    private Task taskFromStringCsv(String value) {
+        if (value == null || value.isEmpty()) return null;
+
+        String[] parts = value.split(SEPARATOR_CSV);
+        if (parts.length < 5) return null; // нарушена структура строки
+
+        TypeTask type;
+        int id;
+        StatusTask status;
+        Task task;
+
+        try {
+            type = TypeTask.valueOf(parts[1]);
+            id = Integer.parseInt(parts[0]);
+            status = StatusTask.valueOf(parts[3]);
+
+            task = switch (type) {
+                case TypeTask.TASK -> new Task(parts[2], parts[4], id, status);
+                case TypeTask.EPIC_TASK -> new EpicTask(parts[2], parts[4], id, status, new ArrayList<>());
+                case TypeTask.SUB_TASK -> new SubTask(parts[2], parts[4], id, status, Integer.parseInt(parts[5]));
+            };
+        } catch (IllegalArgumentException e) {
+            task = null;
+        }
+        return task;
+    }
```
Это методы бы вынести в отдельный утилитный класс, чтобы разгрузить файловый менеджер и оставить в нем только логику сохранения/чтения файла и переопределенные методы менеджера

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,166 @@
+
+// Решил что логика преобразования в CSV необходима только в данном классе, поэтому не встраивал в классы Task
+
+public class FileBackedTaskManager extends InMemoryTaskManager implements TaskManager {
```
Имплементить интерфейс `TaskManager` не нужно, он и так уже имплементируется классом `InMemoryTaskManager`

ds-skoropad replied:
Это фаталити :)

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,166 @@
+        setHistoryManager(historyManager);
+    }
+
+    public Path getFile() {
+        return file;
+    }
+
+    public void setFile(Path file) {
+        this.file = file;
+    }
```
Не добавляй лишние методы, если они не используются)

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,166 @@
+                case TypeTask.EPIC_TASK -> new EpicTask(parts[2], parts[4], id, status, new ArrayList<>());
+                case TypeTask.SUB_TASK -> new SubTask(parts[2], parts[4], id, status, Integer.parseInt(parts[5]));
+            };
+        } catch (IllegalArgumentException e) {
```
Не лучшая практика обрабатывать непроверяемые исключения, их можно избежать кодом.

ds-skoropad replied:
Действительно, в курсе было сказано не увлекаться исключениями и обходить кодом. Благодарю за замечание!

Ivanov-s-tmn commented on test/ru/yandex/practicum/kanban/manager/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,81 @@
+        assertTrue(fb.getSubTaskGroup().isEmpty());
+    }
+
+    @Test
+    void save() throws IOException {
+        fb.addTask(new Task("Task 1", "Description"));
+        fb.addTask(new EpicTask("Epic 1", "Description"));
+        fb.addTask(new SubTask("Sub 1", "Description", 3, StatusTask.NEW, 2));
+
+        ArrayList<String> data = new ArrayList<>();
+        BufferedReader reader = new BufferedReader(new FileReader(tempFile.toFile(), StandardCharsets.UTF_8));
+        while (reader.ready()) {
+            data.add(reader.readLine());
+        }
+        reader.close();
+
+        assertArrayEquals(data.toArray(), fileContent);
+    }
+
+    @Test
+    void load() throws IOException {
+        BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile.toFile(), StandardCharsets.UTF_8));
+        for (String s : fileContent) {
+            writer.write(s);
+            writer.newLine();
+        }
+        writer.close();
+
+        fb.loadFromFile(tempFile);
+
+        assertNotNull(fb.getTask(1));
+        assertNotNull(fb.getEpicTask(2));
+        assertNotNull(fb.getSubTask(3));
+    }
```
Самое главное в тестах файлового менеджера проверить корректность восстановления состояния из файла, тест кейс будет примерно таким:
Сохранение задач в файловый менеджер работает при вызове методов сохранения/удаления/обновления, так как эта логика вшита в его переопределенные методы таск менеджера. Соответственно в тестах для создания, удаления и обновления задач нужно пользоваться только методами файлового менеджера, а выгружать данные из файла с помощью `loadFromFile()`, как пример тест-кейса, который нужно реализовать:

1. Создать файловый менеджер
2. Заполнить тестовыми данными
3. Создать новый файловый менеджер из этого же файла через loadFromFile()
4. Проверить, что новый менеджер корректно восстановил состояние из файла (списки задач до и после выгрузки должны быть идентичны)
Этот один тест будет проверять как запись в файл, так и восстановление данных из файла

ds-skoropad committed:
refactor: rebuild FileBackedTaskManager

* ManagerSaveException separated
* Added TaskUtils class fo CSV methods
* Loading from a file into task group maps
* Change access maps (InMemoryTaskManager) for loadFromFile()
Some extra fix.

Sprint 7 Review 2

ds-skoropad committed:
style(util): give space before switch

Added a "data" variable to allow the "switch" to go to a new line.

Sprint 7 Review 2

ds-skoropad commented:
**TaskUtil**
Реализация заголовка файла csv через enum, что позволило отказаться от непонятных индексов при формировании строк\объектов задач и улучшить чтение кода.
Также можно менять порядок полей в заголоке изменяя порядок в enum не затрагивая изменения кода методов. Добавить новые поля заголовка при желании тоже будет проще.

**Коммит** [[e5278b9](https://github.com/ds-skoropad/java-kanban/pull/4/commits/e5278b97a83966c36436774b153baef5a7ee544b)]
Произведена правка из-за ошибки отсутствия отступа перед оператором switch. При проверке по чек-стилям от Google и Sun данная ошибка не выявлена. 
Добавление пробела перед switch в данном случае или перевод на новую строку не будет соответствовать стилю кода внутри курса. 
Решено через добавление переменной.
Похожая проблема https://github.com/checkstyle/checkstyle/issues/14825.

ds-skoropad commented on src/ru/yandex/practicum/kanban/util/TaskUtils.java:
```diff
@@ -0,0 +1,107 @@
+public final class TaskUtils {
+    private static final String CSV_SEPARATOR = ",";
+
+    private enum CsvHead {
```
Реализация заголовка файла csv через enum, что позволило отказаться от непонятных индексов при формировании строк\объектов задач и улучшить чтение кода.

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/util/TaskUtils.java:
```diff
@@ -0,0 +1,109 @@
+        if (stringInArrEnum(TypeTask.values(), parts[CsvHead.type.ordinal()])) {
+            type = TypeTask.valueOf(parts[CsvHead.type.ordinal()]);
+        } else {
+            return null;
```
Не лучшая практика возвращать явно null из методов, тут можно сделать иначе, к примеру инвертировать условие 
```java
if (!stringInArrEnum(TypeTask.values(), parts[CsvHead.type.ordinal()])) {
    //выбросить исключение с сообщением об ошибке
}
TypeTask typetype = TypeTask.valueOf(parts[CsvHead.type.ordinal()]);
```
Так не будет возвратов null, плюс, не будет else блоков и сэкономим несколько строк

Ivanov-s-tmn commented on test/ru/yandex/practicum/kanban/manager/FileBackedTaskManagerTest.java:
```diff
@@ -1,81 +1,39 @@
-            data.add(reader.readLine());
-        }
-        reader.close();
+        assertArrayEquals(fb1.getTaskGroup().toArray(), fb2.getTaskGroup().toArray());
+        assertArrayEquals(fb1.getSubTaskGroup().toArray(), fb2.getSubTaskGroup().toArray());
```
Можно сравнивать и просто списки через `assertEquals()`

Ivanov-s-tmn commented on test/ru/yandex/practicum/kanban/manager/FileBackedTaskManagerTest.java:
```diff
@@ -1,81 +1,39 @@
-        assertNotNull(fb.getTask(1));
-        assertNotNull(fb.getEpicTask(2));
-        assertNotNull(fb.getSubTask(3));
+        Files.deleteIfExists(tempFile);
```
Создание и удаление временного файла будет лучше вынести в BeforeEach/AfterEach методы. Было бы неплохо проверить загрузку на пустом файле, все должно отрабатывать без ошибок

Ivanov-s-tmn commented on test/ru/yandex/practicum/kanban/manager/FileBackedTaskManagerTest.java:
```diff
@@ -1,81 +1,39 @@
+        assertEquals(epic1.size(), epic2.size());
+        for (int i = 0; i < epic1.size(); i++) {
+            assertEquals(epic1.get(i), epic2.get(i));
+            assertArrayEquals(epic1.get(i).getSubTaskIds().toArray(), epic2.get(i).getSubTaskIds().toArray());
```
Циклов в тестах стоит избегать, просто сделай проверку списка эпиков, как задач и подзадач, а потом отдельно сравни список айди подзадач внутри эпика:
`assertEquals(epic1.getSubTaskIds(), epic2.getSitasksIds());`

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -83,84 +74,69 @@ public boolean removeEpicTask(int id) {
+                                if (fb.epicTaskGroup.containsKey(sub.getEpicTaskId())) {
+                                    fb.subTaskGroup.put(sub.getId(), sub);
+                                    fb.epicTaskGroup.get(sub.getEpicTaskId()).addSubTaskId(sub.getId());
+                                    if (task.getId() > idMax) idMax = task.getId();
```
Эти проверки дублируются и ее можно по идее делать в начале цикла до свитча

ds-skoropad committed:
refactor(file-backed): don't return null

Sprint 7 Review 3

ds-skoropad committed:
refactor(file-backed): Clear the code

Sprint 7 Review 3
